File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/Python/SingleNumber.py
# Given an array of integers, every element appears twice except for one. Find that single one.
# 
# Note:
# Your algorithm should have a linear runtime complexity. Could you implement it without using extra memory? 

class Solution:
    # @param A, a list of integer
    # @return an integer
    def singleNumber(self, A):
        # XOR each number, the final number is the one only appears once
        
        # reduce(function, iterablt[, initializer])
        #   apply function of two arguments cumulatively to the items of iterable

        # operator.xor(a, b)  return the bitwise exclusive or of a and b
        return reduce(operator.xor, A)

File: SQL-AI/ProgrammingLanguage/Leetcode-Implementation/Ruby/single_number.rb
#  Given an array of integers, every element appears twice except for one. Find that single one.
#
#  Note:
#  Your algorithm should have a linear runtime complexity. Could you implement it without using extra memory?

# @param {Integer[]} nums
# @return {Integer}
def single_number(nums)
  # bit maniipulation XOR
  single = 0
  for num in nums
    single = single ^ num
  end
  return single
end

# Hash table
def single_number_1(nums)
  hash = Hash.new
  for num in nums
    if !hash[num]
      hash[num] = true
    else
      hash.delete(num)
    end
  end
  puts hash.keys[0]
end

single_number_1([1,2,2,1,44,4,4,55,55,44,9])
